src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Threading
+{
+    internal static partial class WaitSubsystem
+    {
+        [EagerStaticClassConstruction] // used during lazy class construction
+        private static class HandleManager
+        {
+            private const int InitialCapacity = 4;
+
+            /// <summary>
+            /// Prevents reusing a deleted index until at least there are this many deleted indexes in the free list. The
+            /// purpose is to allow free handles to remain free for some time so that we can discover invalid attempts at using
+            /// free handles.
+            /// </summary>
+            private const int MinimumFreeIndexCountBeforeReuse = 64;
+
+            private static WaitableObject[] s_objects = new WaitableObject[InitialCapacity];
+            private static int s_count;
+            private static LowLevelQueue<int> s_freeIndexes;
```

Comment:
> LowLevelList

And LowLevelListWithIList, LowLevelDictionary, LowLevelDictionaryWithIEnumerable, ConcurrentUnifier, ConcurrentUnifierW, ConcurrentUnifierWKeyed, LowLevelListDictionary, ... there many broken collection clones. That's why I am not thrilled about adding another one for what seems pretty low value. I would not be opposed to adding one when it is justified.

I believe that something like handle manager is best written directly against metal and kept as simple possible.

From what I can tell, CoreCLR PAL handle manager has a simple free list: https://github.com/dotnet/coreclr/blob/32f0f9721afb584b4a14d69135bea7ddc129f755/src/pal/src/handlemgr/handlemgr.cpp#L145. I do not remember anybody ever wishing for having a more elaborate one, and the risk for screw-ups in CoreCLR is higher because of it is a ton of C++. I am wondering why we cannot have just that here as well.